### PR TITLE
Added CMake variables to select the modules to be built

### PR DESCRIPTION
**Pull request by MarioLiebisch**
This addresses issue #798.

The only leftover issue I see is the fact that on most platforms we're always installing most third party dependencies, even if the modules aren't built.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -44,11 +44,25 @@ else()
     set(SFML_BUILD_EXAMPLES FALSE)
 endif()
 
+# add options to select which modules to build
+sfml_set_option(SFML_BUILD_SYSTEM TRUE BOOL "TRUE to build SFML's System module. This setting is ignored, if any of the other modules is built.")
```
Comment (LaurentGomila) [inline CMakeLists.txt]: What's the point of not building sfml-system? I think we can remove this option.

File: CMakeLists.txt [hunk 2]
```diff
@@ -44,11 +44,25 @@ else()
...
+if(SFML_OS_WINDOWS OR SFML_OS_ANDROID OR SFML_OS_IOS)
+    sfml_set_option(SFML_BUILD_MAIN TRUE BOOL "TRUE to build SFML's main module (entry point handling).")
```
Comment (LaurentGomila) [inline CMakeLists.txt]: I don't think we need this option either. On iOS and Android it is mandatory, and on Windows it's almost always wanted. And it's just a single C++ file compiled into a static library.

File: CMakeLists.txt [hunk 3]
```diff
@@ -44,11 +44,25 @@ else()
...
+    sfml_set_option(SFML_BUILD_AUDIO TRUE BOOL "TRUE to build SFML's Audio module.")
+endif()
+sfml_set_option(SFML_BUILD_NETWORK TRUE BOOL "TRUE to build SFML's Network module.")
+
 # add an option for building the API documentation
 sfml_set_option(SFML_BUILD_DOC FALSE BOOL "TRUE to generate the API documentation, FALSE to ignore it")
 
 # add an option for choosing the OpenGL implementation
+if(SFML_BUILD_GRAPHICS OR SFML_BUILD_WINDOW)
 sfml_set_option(SFML_OPENGL_ES ${OPENGL_ES} BOOL "TRUE to use an OpenGL ES implementation, FALSE to use a desktop OpenGL implementation")
```
Comment (LaurentGomila) [inline CMakeLists.txt]: Indentation :smile_cat:

**Comment by LaurentGomila** [comment]
What's the problem with external libs? Your PR seems to handle them correctly.

File: CMakeLists.txt (hunk 3 above)
Comment (MarioLiebisch) [inline CMakeLists.txt]: RAWRARWRARAR RAGE!

File: CMakeLists.txt (hunk 1 above)
Comment (MarioLiebisch) [inline CMakeLists.txt]: Maybe building documentation only? Wasn't sure either.

Comment (mantognini) [inline CMakeLists.txt]: We already can build only the doc with `make doc`. ;-)

Comment (MarioLiebisch) [inline CMakeLists.txt]: It would still want dependencies and you couldn't install the docs only. But if you think we don't need that option, can cut it as well (same with the main one).

Comment (mantognini) [inline CMakeLists.txt]: Indeed, I forget the issue at hand, sorry.

But wouldn't someone who's building the doc want to build at least one module of SFML?

**Comment by MarioLiebisch** [comment]
No problem in general, but let's assume you want to install system and network only (server), you'd still install all extlibs.

File: CMakeLists.txt (hunk 1 above)
Comment (MarioLiebisch) [inline CMakeLists.txt]: Yeah, it's quite an edge case.

**Comment by eXpl0it3r** [comment]
> you'd still install all extlibs (i.e. on Windows)

Did you test this? Afaik the extlibs have to be explicitly added to get installed so when you just pick one module it won't added the other ones, but maybe my memory serves me wrong.

**Comment by MarioLiebisch** [comment]
Right now you install the whole directory since the file names are toolchain dependent.

**Comment by MarioLiebisch** [comment]
So, any further comments regarding this? Especially related to whether to include `SFML_BUILD_SYSTEM` or ommit it.

**Comment by mantognini** [comment]
I also believe we don't need `SFML_BUILD_SYSTEM`. Beside that, this PR's changes will conflict with change with no_libsndfile but that's not a big deal.

**Comment by binary1248** [comment]
This PR needs rebasing now that no_libsndfile is merged.

**Comment by MarioLiebisch** [comment]
Completely forgot about this; the branch is rebased and up to date again. Still not 100% sold regarding `SFML_BUILD_SYSTEM` though.

**Comment by Bromeon** [comment]
I would remove the option `SFML_BUILD_SYSTEM`, too. You just add confusion and possible error sources with more options. For example, you would then need to do all kinds of checks like "you can't use sfml-audio without sfml-system".

**Comment by MarioLiebisch** [comment]
> For example, you would then need to do all kinds of checks like "you can't use sfml-audio without sfml-system".

To prevent this, dependencies are always built, so even with `SFML_BUILD_SYSTEM` set to false it's always built in case you try to build any of the other libraries.

But yeah, guess I'll just drop it then, unless there's some other opinion till tomorrow.

**Comment by Bromeon** [comment]
> To prevent this, dependencies are always built, so even with `SFML_BUILD_SYSTEM` set to false it's always built in case you try to build any of the other libraries.

I personally don't like this way of hiding user errors. If the user ticks sfml-graphics but not sfml-system, he probably has a misconception of how the dependencies work -- and I'd rather tell him than just assuming he didn't mean it. Same with sfml-graphics and sfml-window at the moment, I would probably add an error message...

**Comment by MarioLiebisch** [comment]
True, there should be a warning, if it's possible to exclude dependencies, which are still built.

**Comment by mantognini** [comment]
So what's the status of this?

In any case, [those lines](https://github.com/SFML/SFML/blob/feature/selective_building/CMakeLists.txt#L331-L345) need to be updated.

**Comment by MarioLiebisch** [comment]
Can have a look later this week; caught a cold. :(

**Comment by mantognini** [comment]
Before I forget: the framework part for OS X also need some update. See https://github.com/SFML/SFML/pull/933/files#diff-af3b638bc2a3e6c650974192a53c7291R272 (not tested)

**Comment by MarioLiebisch** [comment]
So we have any current open PR or changes that would conflict this? I'd rebase it, but don't want to do the work twice. :)

**Comment by Bromeon** [comment]
Are we going to include this in SFML 2.4? If so, it would be nice to perform the adaptations mentioned above.

**Comment by eXpl0it3r** [comment]
@MarioLiebisch Can you rebase this and include the possible changes that @mantognini mentioned?

**Comment by MarioLiebisch** [comment]
Done.

File: CMakeLists.txt [hunk 4]
```diff
@@ -329,31 +339,33 @@ if(SFML_OS_WINDOWS)
 elseif(SFML_OS_MACOSX)
 
     # install extlibs dependencies only when used
-    if("${FLAC_LIBRARY}" STREQUAL "${SFML_SOURCE_DIR}/extlibs/libs-osx/Frameworks/FLAC.framework")
+    if(SFML_BUILD_AUDIO AND FLAC_LIBRARY STREQUAL "${SFML_SOURCE_DIR}/extlibs/libs-osx/Frameworks/FLAC.framework")
```
Comment (mantognini) [inline CMakeLists.txt]: Audio dependencies should be moved together in a single `if(SFML_BUILD_AUDIO)` block. This includes Flac,

File: CMakeLists.txt [hunk 5]
```diff
@@ -329,31 +339,33 @@ if(SFML_OS_WINDOWS)
...
-    if("${FREETYPE_LIBRARY}" STREQUAL "${SFML_SOURCE_DIR}/extlibs/libs-osx/Frameworks/freetype.framework")
-        install(DIRECTORY extlibs/libs-osx/Frameworks/freetype.framework DESTINATION ${CMAKE_INSTALL_FRAMEWORK_PREFIX})
+    if(SFML_BUILD_GRAPHICS)
+        if(FREETYPE_LIBRARY STREQUAL "${SFML_SOURCE_DIR}/extlibs/libs-osx/Frameworks/freetype.framework")
+            install(DIRECTORY extlibs/libs-osx/Frameworks/freetype.framework DESTINATION ${CMAKE_INSTALL_FRAMEWORK_PREFIX})
+        endif()
     endif()
 
-    if("${OGG_LIBRARY}" STREQUAL "${SFML_SOURCE_DIR}/extlibs/libs-osx/Frameworks/ogg.framework")
+    if(OGG_LIBRARY STREQUAL "${SFML_SOURCE_DIR}/extlibs/libs-osx/Frameworks/ogg.framework")
```
Comment (mantognini) [inline CMakeLists.txt]: ogg,

File: CMakeLists.txt [hunk 6]
```diff
@@ -329,31 +339,33 @@ if(SFML_OS_WINDOWS)
...
         install(DIRECTORY extlibs/libs-osx/Frameworks/ogg.framework DESTINATION ${CMAKE_INSTALL_FRAMEWORK_PREFIX})
     endif()
 
-    if("${VORBIS_LIBRARY}" STREQUAL "${SFML_SOURCE_DIR}/extlibs/libs-osx/Frameworks/vorbis.framework")
+    if(VORBIS_LIBRARY STREQUAL "${SFML_SOURCE_DIR}/extlibs/libs-osx/Frameworks/vorbis.framework")
```
Comment (mantognini) [inline CMakeLists.txt]: vorbis,

File: CMakeLists.txt [hunk 7]
```diff
@@ -329,31 +339,33 @@ if(SFML_OS_WINDOWS)
...
         install(DIRECTORY extlibs/libs-osx/Frameworks/vorbis.framework DESTINATION ${CMAKE_INSTALL_FRAMEWORK_PREFIX})
     endif()
 
-    if("${VORBISENC_LIBRARY}" STREQUAL "${SFML_SOURCE_DIR}/extlibs/libs-osx/Frameworks/vorbisenc.framework")
+    if(VORBISENC_LIBRARY STREQUAL "${SFML_SOURCE_DIR}/extlibs/libs-osx/Frameworks/vorbisenc.framework")
```
Comment (mantognini) [inline CMakeLists.txt]: vorbisenc,

File: CMakeLists.txt [hunk 8]
```diff
@@ -329,31 +339,33 @@ if(SFML_OS_WINDOWS)
...
         install(DIRECTORY extlibs/libs-osx/Frameworks/vorbisenc.framework DESTINATION ${CMAKE_INSTALL_FRAMEWORK_PREFIX})
     endif()
 
-    if("${VORBISFILE_LIBRARY}" STREQUAL "${SFML_SOURCE_DIR}/extlibs/libs-osx/Frameworks/vorbisfile.framework")
+    if(VORBISFILE_LIBRARY STREQUAL "${SFML_SOURCE_DIR}/extlibs/libs-osx/Frameworks/vorbisfile.framework")
```
Comment (mantognini) [inline CMakeLists.txt]: vorbisfile,

File: CMakeLists.txt [hunk 9]
```diff
@@ -329,31 +339,33 @@ if(SFML_OS_WINDOWS)
...
         install(DIRECTORY extlibs/libs-osx/Frameworks/vorbisfile.framework DESTINATION ${CMAKE_INSTALL_FRAMEWORK_PREFIX})
     endif()
 
-    if("${OPENAL_LIBRARY}" STREQUAL "${SFML_SOURCE_DIR}/extlibs/libs-osx/Frameworks/OpenAL.framework")
+    if(OPENAL_LIBRARY STREQUAL "${SFML_SOURCE_DIR}/extlibs/libs-osx/Frameworks/OpenAL.framework")
```
Comment (mantognini) [inline CMakeLists.txt]: and OpenAL.

**Comment by mantognini** [comment]
Thanks for the rebase BTW. 👍

File: CMakeLists.txt (hunk 6 above)
Comment (MarioLiebisch) [inline CMakeLists.txt]: Updated, although I think these should ultimatively moved to the modules' CMakeLists.txt rather than doing it globally here.

**Comment by eXpl0it3r** [comment]
Looks like this PR was a bit forgotten. Can you give this a rebase?

**Comment by eXpl0it3r** [comment]
Nearly feel bad for asking again, can you give this one last rebase? 😄
